server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -320,39 +319,15 @@ String[] concreteIndexNames(Context context, String... indexExpressions) {
     }
 
     Index[] concreteIndices(Context context, String... indexExpressions) {
-        IndicesOptions options = context.getOptions();
-        if (indexExpressions == null || indexExpressions.length == 0) {
-            indexExpressions = new String[] { Metadata.ALL };
-        } else {
-            if (options.ignoreUnavailable() == false) {
-                List<String> crossClusterIndices = Arrays.stream(indexExpressions).filter(index -> index.contains(":")).toList();
-                if (crossClusterIndices.size() > 0) {
-                    throw new IllegalArgumentException(
-                        "Cross-cluster calls are not supported in this context but remote indices "
-                            + "were requested: "
-                            + crossClusterIndices
-                    );
-                }
-            }
-        }
+        ensureRemoteIndicesRequireIgnoreUnavailable(context.getOptions(), indexExpressions);
+        final Collection<String> expressions = resolveExpressions(context, indexExpressions);
 
-        final Collection<String> expressions = resolveExpressions(Arrays.asList(indexExpressions), context);
-
-        if (expressions.isEmpty() || (expressions.size() == 1 && expressions.iterator().next().equals(Metadata.ALL))) {
-            if (options.allowNoIndices() == false) {
-                throw notFoundException(indexExpressions);
-            } else {
-                return Index.EMPTY_ARRAY;
-            }
-        }
-
-        final Set<Index> concreteIndices = Sets.newLinkedHashSetWithExpectedSize(expressions.size());
-        final SortedMap<String, IndexAbstraction> indicesLookup = context.state.metadata().getIndicesLookup();
+        final Set<Index> concreteIndicesResult = Sets.newLinkedHashSetWithExpectedSize(expressions.size());
         for (String expression : expressions) {
-            if (options.ignoreUnavailable() == false) {
+            if (context.getOptions().ignoreUnavailable() == false) {
                 ensureAliasOrIndexExists(context, expression);
             }
-            IndexAbstraction indexAbstraction = indicesLookup.get(expression);
+            final IndexAbstraction indexAbstraction = context.state.metadata().getIndicesLookup().get(expression);
```

Comment:
What's the reasoning for doing this multi-stage dereference in every loop?

Reply:
Pushed [c588555](https://github.com/elastic/elasticsearch/pull/91735/commits/c58855555cd8c1a728b2f3a509835662a313d79a) .